### PR TITLE
fix(controller): Report reconciliation errors better

### PR DESCRIPTION
**Pull request by alexec**
Signed-off-by: Alex Collins <alex_collins@intuit.com>

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

**Comment by alexec** [comment]
`TestContinueOnFailDag`

File: .github/workflows/ci-build.yaml [hunk 1]
```diff
@@ -111,7 +111,8 @@ jobs:
           if [ ${{matrix.test}} == smoke ] || [ ${{matrix.test}} == test-e2e-cron ] ; then
             PROFILE=minimal
           fi
-          KUBECONFIG=~/.kube/config make start PROFILE=$PROFILE E2E_EXECUTOR=${{matrix.containerRuntimeExecutor}} ALWAYS_OFFLOAD_NODE_STATUS=${{matrix.alwaysOffloadNodeStatus}} DEV_IMAGE=true STATIC_FILES=false 2>&1 > /tmp/log/argo-e2e/argo.log &
+          KUBECONFIG=~/.kube/config make install PROFILE=$PROFILE E2E_EXECUTOR=${{matrix.containerRuntimeExecutor}} ALWAYS_OFFLOAD_NODE_STATUS=${{matrix.alwaysOffloadNodeStatus}} DEV_IMAGE=true STATIC_FILES=false
+          KUBECONFIG=~/.kube/config make start   PROFILE=$PROFILE E2E_EXECUTOR=${{matrix.containerRuntimeExecutor}} ALWAYS_OFFLOAD_NODE_STATUS=${{matrix.alwaysOffloadNodeStatus}} DEV_IMAGE=true STATIC_FILES=false 2>&1 > /tmp/log/argo-e2e/argo.log &
```
Comment (alexec) [inline .github/workflows/ci-build.yaml:115]: new PR for changes to this file

File: workflow/controller/controller_test.go [hunk 2]
```diff
@@ -263,6 +263,14 @@ func expectWorkflow(ctx context.Context, controller *WorkflowController, name st
 	test(wf)
 }
 
+func getPod(woc *wfOperationCtx, name string) (*apiv1.Pod, error) {
```
Comment (alexec) [inline workflow/controller/controller_test.go:266]: new PR for these changes

File: workflow/controller/operator.go [hunk 3]
```diff
@@ -344,13 +344,20 @@ func (woc *wfOperationCtx) operate(ctx context.Context) {
...
+		woc.log.WithError(err).Error("error in entry template execution")
+		// we wrap this error up to report a clear message
+		x := fmt.Errorf("error in entry template execution: %w", err)
 		switch err {
 		case ErrDeadlineExceeded:
-			woc.eventRecorder.Event(woc.wf, apiv1.EventTypeWarning, "WorkflowTimedOut", msg)
+			woc.eventRecorder.Event(woc.wf, apiv1.EventTypeWarning, "WorkflowTimedOut", x.Error())
+		case ErrParallelismReached:
+		case ErrTimeout:
+			woc.markWorkflowFailed(ctx, x.Error())
```
Comment (alexec) [inline workflow/controller/operator.go:355]: needs guard

File: workflow/controller/operator.go [hunk 4]
```diff
@@ -344,13 +344,20 @@ func (woc *wfOperationCtx) operate(ctx context.Context) {
...
+			return
+		default:
```
Comment (alexec) [inline workflow/controller/operator.go:357]: guard with transient check

Comment (alexec) [reply workflow/controller/operator.go:357]: add feature flag (envvar)

File: workflow/controller/operator.go [hunk 5]
```diff
@@ -344,13 +344,20 @@ func (woc *wfOperationCtx) operate(ctx context.Context) {
 
 	node, err := woc.executeTemplate(ctx, woc.wf.ObjectMeta.Name, &wfv1.WorkflowStep{Template: woc.execWf.Spec.Entrypoint}, tmplCtx, woc.execWf.Spec.Arguments, &executeTemplateOpts{})
 	if err != nil {
-		// the error are handled in the callee so just log it.
-		msg := "error in entry template execution"
-		woc.log.WithError(err).Error(msg)
-		msg = fmt.Sprintf("%s %s: %+v", woc.wf.Name, msg, err)
+		woc.log.WithError(err).Error("error in entry template execution")
```
Comment (alexec) [inline workflow/controller/operator.go:347]: * [ ] search for this and see what happens

**Comment by alexec** [comment]
```
time="2021-01-16T20:11:13Z" level=error msg="error in entry template execution" error="pods \"canary-56gqx\" is forbidden: rpc error: code = Unavailable desc = transport is closing" namespace=argo workflow=canary-56gqx
clustername = ip-asterias-prod-usw2-k8shost = workflow-controller-79dd88b9f-k8jjvsource = kubernetes.var.log.containers.workflow-controller-79dd88b9f-k8jjv_argo_workflow-controller-cb40482b7a10b9374d7b7d5242c8cc2b69e81ef6e8304de3f5e5c14f1dd09bcc.log
```

File: util/diff/diff.go [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+)
+
+func LogChanges(old, new interface{}) {
+	if !log.IsLevelEnabled(log.DebugLevel) {
+		return
+	}
+	a, _ := json.Marshal(old)
+	b, _ := json.Marshal(new)
+	patch, _ := jsonpatch.CreateMergePatch(a, b)
+	log.Debug(string(patch))
```
Comment (simster7) [inline util/diff/diff.go:17]: If we're going to keep debug statements, can we place add an identifier string to them? E.g.
```suggestion
	log.Debugf("Log changes patch: %s", string(patch))
```
This way if it shows up in the logs its easy to find

Comment (simster7) [reply util/diff/diff.go:17]: Also, very minor, but I would consider having this function to return the string and for the caller of the function to do the logging.

File: workflow/controller/operator.go [hunk 7]
```diff
@@ -344,13 +345,17 @@ func (woc *wfOperationCtx) operate(ctx context.Context) {
...
+		// we wrap this error up to report a clear message
+		x := fmt.Errorf("error in entry template execution: %w", err)
 		switch err {
 		case ErrDeadlineExceeded:
-			woc.eventRecorder.Event(woc.wf, apiv1.EventTypeWarning, "WorkflowTimedOut", msg)
+			woc.eventRecorder.Event(woc.wf, apiv1.EventTypeWarning, "WorkflowTimedOut", x.Error())
+		case ErrParallelismReached:
```
Comment (simster7) [inline workflow/controller/operator.go:354]: Currently we won't log `ErrParallelismReached`. Is this intended or did you mean to add a `fallthrough`?

**Review by simster7 — APPROVED** [review]

**Pull request merged**